### PR TITLE
Fix/10731 auto focus input

### PR DESCRIPTION
**Pull request by zzz0108**
### Fixed Issues

$ https://github.com/Expensify/App/issues/10731
PROPOSAL: https://github.com/Expensify/App/issues/10731#issuecomment-1255955313

### Tests

* 10731

1. Go to any chats
2. Verify that the composer is focused and the virtual keyboard is not shown. After touching on the composer, the keyboard is shown
3. Sent any message
4. Open edit mode
5. Verify that the keyboard is shown
6. Tap on the composer to bring up context menu.
7. Verify the context menu is shown 

* 50683

1. Login to an account
2. Open any chat, notice that the composer is focused.
3. Press on CTRL + K
4. Press ESC key to dismissed the 'search modal'
5. Verify that: Composer is focused after dismissing the modal

* 50722

1. Open the Hybrid app/NewDot app and log in as user A
2. On a secondary device as user B send an IOU to user A
3. On main device as user A open the chat with user B
4. Select a different payment option on the IOU preview
5. Verify that: The page does not flicker when changing a payment option

- [x] Verify that no errors appear in the JS console

### Offline tests

Same as above

### QA Steps

Same as above

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/f8324505-c657-4c1c-bd10-ba7e48ac032d

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/7623eae1-79ac-43d6-886e-58a0c509d9e2

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/08c133d2-35d8-415d-9359-15f7e3b374f9

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/d5a820b7-7211-4c35-8f16-a63398fb174b

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/61dafb30-a167-4bce-96cf-ac6d31a53dd0

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/7acf9186-5c5a-4188-a804-b2807971aa2e

</details>

### CODE REVIEW

**Comment by DylanDylann** [comment]
Reviewing today

File: src/components/Composer/implementation/index.tsx [hunk 1]
```diff
@@ -281,24 +281,28 @@ function Composer(
         onClear(currentText);
     }, [onClear, onSelectionChange]);
 
-    useImperativeHandle(ref, () => {
-        const textInputRef = textInput.current;
-        if (!textInputRef) {
-            throw new Error('textInputRef is not available. This should never happen and indicates a developer error.');
-        }
+    useImperativeHandle(
```
Comment (DylanDylann) [inline src/components/Composer/implementation/index.tsx:284]: @christianwen  Is there any new update on this file?

**Comment by DylanDylann** [comment]
@christianwen Could you also please help to merge the latest main?

**Comment by DylanDylann** [comment]
@christianwen It is a bit flicker on Safari

https://github.com/user-attachments/assets/978d2791-59f8-4da5-bfd9-6e6104baf5c4

**Comment by DylanDylann** [comment]
@christianwen BUG: when going to the chat, the cursor displays in the middle or in the end of the composer in a moment 
Step:
1. Go to chat A (Don't see the bug)
2. Back to LHN
3. Go to chat A (See the Bug)
4. Back to LHN
5. Go to chat A (See the Bug)

https://github.com/user-attachments/assets/a51624df-d6e2-4400-8750-add2618b48cf

**Comment by DylanDylann** [comment]
@christianwen BUG
1. Focus on the edit composer
2. Click the dropdown button or pay button 
3. The main composer gets focused again

On staging: The main composer isn't getting focused again

https://github.com/user-attachments/assets/16591745-b8f9-4bf7-ba5b-47161b1c85bf

**Comment by DylanDylann** [comment]
BUG: The composer is hidden by the keyboard

Behavior on this change

https://github.com/user-attachments/assets/d0c994a1-8d5e-4e0c-bd09-803db27a28ec



Behavior on the main 

https://github.com/user-attachments/assets/215b8c0f-7615-40bd-b600-ece2c214c16f

**Comment by DylanDylann** [comment]
> For https://github.com/Expensify/App/pull/51172#issuecomment-2499577331, the RCA is because on small screen, we hide the main composer when focus on edit composer. When users click outside to blur, the main composer will appear and trigger the logic autoFocus (this value is true). I know it's quite different from staging/main but I don't see any problems about this behavior so we can ignore it for now. Pls let me know what you think.

But the cursor is moving to the beginning of main composer and I think it is incorrect.

**Comment by DylanDylann** [comment]
@christianwen 

https://github.com/user-attachments/assets/af9506f6-18c2-443f-a78a-e578b2f5a965

**Comment by DylanDylann** [comment]
Flicker on IOS

1. Go to any chat that have draft message
2. Edit message
3. Click X


https://github.com/user-attachments/assets/3fd0b053-5776-4a90-941c-8368cf7115a2

**Comment by DylanDylann** [comment]
@christianwen Kindly bump

**Comment by DylanDylann** [comment]
@christianwen The change looks fine. Could you please merge the latest main?

**Comment by DylanDylann** [comment]
@christianwen For the failed Eslint, could you also fix it?

**Comment by DylanDylann** [comment]
Same here: https://github.com/Expensify/App/pull/32711#issuecomment-2374541170

> Just a note:
> Reassure Performance Tests
> As mentioned [above](https://github.com/Expensify/App/pull/32711#issuecomment-2363066863)
> 
> > Updated! For the perf issue, we shouldn't mind about it since this test is to compare the number of re-render when users interact with input in baseline (from main) with the current behavior (this PR). In this PR we try to open the keyboard when they focus on the composer -> it takes 2 times to re-render
> 
> This PR will change the current behavior on purpose

**Review by DylanDylann — APPROVED** [review]
LGTM

**Comment by DylanDylann** [comment]
@iwiznia All yours

File: src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.tsx [hunk 2]
```diff
@@ -126,27 +125,26 @@ type ComposerWithSuggestionsProps = Partial<ChildrenProps> & {
...
-    parentReportID: string | undefined;
-
     /** Whether report is from group policy */
     isGroupPolicyReport: boolean;
 
     /** policy ID of the report */
-    policyID: string;
+    policyID?: string;
+
+    /** Whether the soft keyboard is open */
```
Comment (iwiznia) [inline src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.tsx:140]: Isn't this description wrong? Shouldn't it be something like: `Wether to show the keyboard on focus`? (applies everywhere this prop exists)

**Review by iwiznia — APPROVED** [review]

**Comment by iwiznia** [comment]
Hmmm performance tests are failing

**Comment by kirillzyusko** [comment]
I believe this PR is a cause for e2e failures 😔 After merging this PR e2e workflow can't be finished.

File: src/components/Composer/implementation/index.native.tsx [hunk 3]
```diff
@@ -158,6 +171,8 @@ function Composer(
                 props?.onBlur?.(e);
             }}
             onClear={onClear}
+            showSoftInputOnFocus={showSoftInputOnFocus}
+            contextMenuHidden={contextMenuHidden}
```
Comment (mollfpr) [inline src/components/Composer/implementation/index.native.tsx:175]: Hi ✋ Coming from https://github.com/Expensify/App/issues/50682

It's expected that the context menu will show on Android when long presses it. So we think it has different behavior between Android and iOS, and we solve it with using `contextMenuHidden` on iOS only.

**Pull request merged**